### PR TITLE
Bug/invalid sequence does not throw

### PR DESCRIPTION
**Pull request by aljones15**
1. Added eslint-config-digitalbazaar to project
2. when the sequence does not match on update we now throw
3. added a test for it.
4. updated change log with issue addressed.
5. customValidator now blocks attempt to patch on the any id field
https://github.com/digitalbazaar/bedrock-account/issues/7

### CODE REVIEW

File: lib/index.js [hunk 1]
```diff
@@ -15,7 +15,8 @@ const brPermissionCheck = promisify(brPermission.checkPermission);
 const database = require('bedrock-mongodb');
 const jsonpatch = require('fast-json-patch');
 const validateCapabilityDelegation = promisify(
-  brIdentity.validateCapabilityDelegation);
+  brIdentity.validateCapabilityDelegation
+);
```
Comment (dlongley) [inline lib/index.js]: lint rules shouldn't make us do this

File: lib/index.js [hunk 2]
```diff
@@ -322,17 +325,28 @@ api.update = brCallbackify(async ({actor, id, patch, sequence}) => {
...
+    throw new BedrockError(
       'Could not update Account. Record sequence does not match.',
       'InvalidStateError', {
         httpStatusCode: 409,
         public: true,
         actual: sequence,
         expected: record.meta.sequence
-      });
+      }
+    );
```
Comment (dlongley) [inline lib/index.js]: lint rules shouldn't make us do this, original is what we want

File: lib/index.js [hunk 3]
```diff
@@ -322,17 +325,28 @@ api.update = brCallbackify(async ({actor, id, patch, sequence}) => {
...
   }
-
-  const errors = jsonpatch.validate(patch, record.account);
+  const customValidate = (operation, index, tree, existingPath) => {
+    jsonpatch.validator(operation, index, tree, existingPath);
+    const pathId = /\/id$/i.test(existingPath);
+    if(pathId) {
+      throw new jsonpatch.JsonPatchError(
```
Comment (dlongley) [inline lib/index.js]: Interesting ... will have to think about this pattern for handling patch validation.

cc: @davidlehn, @mattcollier, thoughts?

File: lib/index.js (hunk 1 above)
Comment (aljones15) [reply lib/index.js]: but the opposite is ok?

File: lib/index.js (hunk 3 above)
Comment (aljones15) [reply lib/index.js]: open to alternatives, but we could look into expanding the json validator with more cases.

File: lib/index.js (hunk 1 above)
Comment (aljones15) [reply lib/index.js]: the real issue is really with callbacks.
when we have callbacks in paramaters they take up multiple lines which results in
odd scenarios. anyways I think they we can get the behavior you want. did not think of this case
line break then no line break when setting up the rule this morning.

File: lib/index.js (hunk 3 above)
Comment (aljones15) [reply lib/index.js]: I do understand the hesitation I mean I am overriding the default validator etc.

Comment (aljones15) [reply lib/index.js]: also might want it to throw a bedrock validator.

File: lib/index.js [hunk 4]
```diff
@@ -322,14 +325,15 @@ api.update = brCallbackify(async ({actor, id, patch, sequence}) => {
   const record = await api.get({actor, id});
 
   if(record.meta.sequence !== sequence) {
-    return new BedrockError(
+    throw new BedrockError(
       'Could not update Account. Record sequence does not match.',
       'InvalidStateError', {
         httpStatusCode: 409,
```
[the remark this answers is not in the document]
Comment (aljones15) [reply lib/index.js]: good catch and thanks for looking at this one will update it in a moment

[the remark this answers is not in the document]
Comment (aljones15) [reply lib/index.js]: actually this is the correct error code:

https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/409

409 is actually for update requests where the data is now invalid.

File: lib/index.js (hunk 2 above)
Comment (aljones15) [reply lib/index.js]: these are the options:

https://eslint.org/docs/rules/function-paren-newline

what is odd is that vue actually does allow the type of behavior you want with html elements inside template so it seems like someone out there would allow line-break then close on the same line as param.

Comment (aljones15) [reply lib/index.js]: might be multiline

File: lib/index.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (dlongley) [reply lib/index.js]: 409 is correct, it's for conflicts (which is what this is).

[the remark this answers is not in the document]
Comment (mattcollier) [reply lib/index.js]: @dlongley what is the current thinking on this?  This seems like more of a SyntaxError(400) than an InvalidStateError.  There's nothing invalid about the state of the server, it's just that bad inputs are being provided to the API.

File: lib/index.js (hunk 1 above)
Comment (dlongley) [reply lib/index.js]: Yes, this should be an error:

```js
promsify(params
) // errors
```

Comment (dlongley) [reply lib/index.js]: > the real issue is really with callbacks.

... and we have a goal of removing callbacks everywhere (eventually) so that should go away or at least be largely mitigated.

File: lib/index.js (hunk 2 above)
Comment (dlongley) [reply lib/index.js]: Seems like we want "never" ... I don't think we have a case where we want to put a parenthesis on its own line by itself.

File: lib/index.js (hunk 3 above)
Comment (dlongley) [reply lib/index.js]: > also might want it to throw a bedrock error.

Well, I think it will fall out and become one so that's perhaps not necessary. I do want feedback from @collier and @davidlehn here ... guys?

File: .eslintrc.js [hunk 5]
```diff
@@ -0,0 +1,7 @@
+module.exports = {
+    "env": {
+        "es6": true,
+        "node": true
+    },
+    "extends": "eslint-config-digitalbazaar",
+};
```
Comment (dlongley) [inline .eslintrc.js]: Again, do we want 2 spaces here like the rest of our JS?

File: lib/index.js [hunk 6]
```diff
@@ -154,8 +155,8 @@ api.insert = brCallbackify(async ({actor, account, meta}) => {
   // validate resource roles (ensure actor is permitted to delegate the
   // roles specified in the account meta)
   if(meta.sysResourceRole.length > 0) {
-    await validateCapabilityDelegation(
-      {actor, resourceRoles: meta.sysResourceRole});
+    const caps = {actor, resourceRoles: meta.sysResourceRole};
+    await validateCapabilityDelegation(caps);
```
Comment (dlongley) [inline lib/index.js]: This is confusing. Why the change?

File: lib/index.js [hunk 7]
```diff
@@ -183,7 +185,8 @@ api.insert = brCallbackify(async ({actor, account, meta}) => {
       'DuplicateError', {
         public: true,
         httpStatusCode: 409
-      }, e);
+      }, e
+    );
```
Comment (dlongley) [inline lib/index.js]: Needs fixing with proper linter rule that never puts parenthesis on its own line by itself.

File: lib/index.js [hunk 8]
```diff
@@ -322,17 +325,28 @@ api.update = brCallbackify(async ({actor, id, patch, sequence}) => {
...
   }
-
-  const errors = jsonpatch.validate(patch, record.account);
+  const customValidate = (operation, index, tree, existingPath) => {
+    jsonpatch.validator(operation, index, tree, existingPath);
+    const pathId = /\/id$/i.test(existingPath);
+    if(pathId) {
+      throw new jsonpatch.JsonPatchError(
+        `You can not change id ${existingPath}`,
```
Comment (dlongley) [inline lib/index.js]: I think we typically don't say "You" in error messages, maybe something like this instead:

`"id" cannot be changed.`

File: lib/index.js (hunk 3 above)
Comment (mattcollier) [reply lib/index.js]: Right, if this customValidator throws, it's going to get caught here and wrapped in a BedrockError right? https://github.com/digitalbazaar/bedrock-account/blob/4d4abc83a5a998a903f0c60149e75609840e5528/lib/index.js#L350-L358

File: lib/index.js (hunk 2 above)
Comment (aljones15) [reply lib/index.js]: I'll set it to never in this project and do a fix and see how it works.

File: lib/index.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (aljones15) [reply lib/index.js]: actually a sequence mismatch means the data on server has been updated while you were making your changes so I think 409 is good here.

File: .eslintrc.js (hunk 5 above)
Comment (aljones15) [reply .eslintrc.js]: I can quickly change this from 4 spaces to 2 no issue.

File: lib/index.js (hunk 6 above)
Comment (aljones15) [reply lib/index.js]: this was causing an eslint parentheses error with line breaks.

File: lib/index.js (hunk 8 above)
Comment (aljones15) [reply lib/index.js]: changing

File: lib/index.js [hunk 9]
```diff
@@ -251,11 +247,10 @@ api.get = brCallbackify(async ({actor, id}) => {
...
   });
 
-  const record = await database.collections.account.findOne(
-    {id: database.hash(id)}, {_id: 0, account: 1, meta: 1});
+  const record = await database.collections.account
+    .findOne({id: database.hash(id)}, {_id: 0, account: 1, meta: 1});
   if(!record) {
-    throw new BedrockError(
-      'Account not found.',
+    throw new BedrockError('Account not found.',
```
Comment (mattcollier) [inline lib/index.js]: If one parameter is indented, all params should be indented (the way it was), this is happening now in several places in the commit.  Revert style of previous code.

File: lib/index.js [hunk 10]
```diff
@@ -593,16 +588,16 @@ api.manageIdentity = brCallbackify(async ({actor, accountId, identityId}) => {
...
   });
 
-  const result = await database.collections.identity.update(
-    {id: database.hash(identityId)},
-    {$set: {'meta.bedrock-account.account': accountId}},
-    // FIXME: call update meta on brIdentity or update sequence here
-    // $inc: {'meta.sequence': 1}
-    database.writeOptions);
+  const result = await database.collections.identity
+    .update({id: database.hash(identityId)},
```
Comment (mattcollier) [inline lib/index.js]: Previous style was correct.

File: lib/index.js (hunk 6 above)
Comment (dlongley) [reply lib/index.js]: We'll want to fix eslint then ... this should be ok and we do it a lot. Also, this was confusing because the variable name was `caps` ... and the above isn't some kind of "capabilities" object, it's two named parameters to a function. If anything I guess it could be called `options`. That being said, I think we should fix eslint so this is ok -- and if there's no option to make it ok, we should disable whatever is making it not ok.

File: lib/index.js [hunk 11]
```diff
@@ -145,17 +145,17 @@ api.insert = brCallbackify(async ({actor, account, meta}) => {
       roles[i] = api.generateResource({role, id: account.id});
     } else if(role.generateResource) {
       // unknown generation directive
-      throw new BedrockError(
-        'Could not create Account; unknown ResourceRole rule.',
+      throw new
+      BedrockError('Could not create Account; unknown ResourceRole rule.',
         'NotSupportedError', {sysResourceRole: role});
```
Comment (dlongley) [inline lib/index.js]: Yeah, this isn't good. We need to figure out the right eslint rules so this is ok -- or if that's not possible, disable whatever is causing this to get messed up.

File: lib/index.js (hunk 9 above)
Comment (dlongley) [reply lib/index.js]: Yeah, needs reverting. Looks like we still don't have the eslint rules quite right.

File: lib/index.js (hunk 2 above)
Comment (aljones15) [reply lib/index.js]: we actually do because with BedRockError some of those strings are so long that

```js
throw new BedrockError(
        'Account does not exist.', 'NotFoundError', {
          httpStatusCode: 404,
          public: true,
        }); // this errors with never and also with consistent

throw new BedrockError( 'Account does not exist and a really long string here.', 
  'NotFoundError', {
     httpStatusCode: 404,
      public: true,
    }); // lint error for line length being over 80
```

what we really need a func-paren-rule that allows
if a parameter's name + function declaration is over 80 then
allow a new line at the start of the declaration. do not allow a new line
at the end of the function paren.

another issue with the never rule is this

```js
  const result = await bd.database.collections.account.findOne( // eslint rule for line length
```

File: lib/index.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (dlongley) [reply lib/index.js]: +1 to what @aljones15 said. That's exactly what's going on. There's no syntax problem from the client -- what they sent is well formed, it's just that the server has moved on and they are out of date.

File: lib/index.js (hunk 6 above)
Comment (aljones15) [reply lib/index.js]: simply put what you want I understand, but the vanilla eslint rule for func-parens just doesn't handle the cases correctly. There has to be a plugin though that does handle it. maybe unicorn has it? we should just remove this rule for the time being until we can find a proper plugin that understands our needs and just make func parens written in an understandable way.

File: lib/index.js (hunk 9 above)
Comment (aljones15) [reply lib/index.js]: yeah this was caused by never with "consistent" if the first paren has a new line the last paren must have a new line. What you want is if paramaters exceed line length then allow a new line otherwise do not.

File: lib/index.js (hunk 10 above)
Comment (aljones15) [reply lib/index.js]: yeah that's what I went with consistent, but paren spacing is really complex.

File: lib/index.js (hunk 11 above)
Comment (aljones15) [reply lib/index.js]: I recommend disable today, reformat to pretty code (or use prettier), then crowd source the right rules.

Comment (aljones15) [reply lib/index.js]: oh yeah we should probably collect a file with all the oddities of our func-paren usage and make sure whatever plugin is choosen it works in every case.

File: lib/index.js (hunk 6 above)
Comment (aljones15) [reply lib/index.js]: we could try: { "minItems": value }
but the issue here is that sometimes we only have 2 paramaters that happen to have really long lengths.

File: lib/index.js (hunk 3 above)
Comment (aljones15) [reply lib/index.js]: is this good to go or not? @gannan08 has not weighed in.

Comment (dlongley) [reply lib/index.js]: I think we're going to accept it for now and reconsider use as we find a need to do more special customization. So we're good in this PR.

File: lib/index.js [hunk 12]
```diff
@@ -359,6 +368,8 @@ api.update = brCallbackify(async ({actor, id, patch, sequence}) => {
   }, database.writeOptions);
 
   if(result.result.n === 0) {
+    // this is what func-paren never set. we might want to define these errors
+    // in their own file so we can reuse.
```
Comment (dlongley) [inline lib/index.js]: Remove comment. I do think we want to figure out a proper rule for this since it was fine before the change, but I'm going to approve this PR for now so we don't keep holding things up.

File: test/mocha/10-api.js [hunk 13]
```diff
@@ -447,6 +449,7 @@ describe('bedrock-account', () => {
         updatedAccount.email = 'UPDATED.' + email;
         const patch = jsonpatch.generate(observer);
         jsonpatch.unobserve(updatedAccount, observer);
+        console.log('updated id', updatedAccount.id);
```
Comment (dlongley) [inline test/mocha/10-api.js]: Remove console.log

File: test/mocha/10-api.js [hunk 14]
```diff
@@ -544,6 +547,116 @@ describe('bedrock-account', () => {
         should.exist(err);
         err.name.should.equal('PermissionDenied');
       });
+      it('should throw Record sequence does not match.', async () => {
```
Comment (dlongley) [inline test/mocha/10-api.js]:
```suggestion
      it('should throw record sequence does not match.', async () => {
```

File: lib/index.js (hunk 12 above)
Comment (dlongley) [reply lib/index.js]: Changed my mind on this... I recommend we just disable the rule that's making this erroneous changes until we get it sorted.

Comment (aljones15) [reply lib/index.js]: done. changes to eslint-project are here: https://github.com/digitalbazaar/eslint-config-digitalbazaar/pull/12

Comment (aljones15) [reply lib/index.js]: oh forgot to remove the comment hold on

File: lib/index.js [hunk 15]
```diff
@@ -331,16 +331,25 @@ api.update = brCallbackify(async ({actor, id, patch, sequence}) => {
...
     throw new BedrockError(
-      'The given JSON patch is invalid.', 'ValidationError', {
+      'The given JSON patch is invalid.',
+      'ValidationError', {
         httpStatusCode: 400,
         public: true,
         patch,
-        errors
-      });
+        errors});
```
Comment (dlongley) [inline lib/index.js]: Let's put this back.

File: lib/index.js [hunk 16]
```diff
@@ -428,8 +437,7 @@ api.updateRoles = brCallbackify(async (
         httpStatusCode: 409,
         public: true,
         actual: sequence,
-        expected: meta.sequence
-      });
+        expected: meta.sequence});
```
Comment (dlongley) [inline lib/index.js]: Put this back

File: lib/index.js [hunk 17]
```diff
@@ -470,9 +478,8 @@ api.updateRoles = brCallbackify(async (
   // 3. ensure resource roles are unique
   const resourceRoles = _.uniqWith(
     brPermission.mergeCapabilities(
-      brPermission.subtractCapabilities(meta.sysResourceRole, remove),
-      add),
-    _.isEqual);
+      brPermission.subtractCapabilities(
+        meta.sysResourceRole, remove), add), _.isEqual);
```
Comment (dlongley) [inline lib/index.js]: Put this back

File: lib/index.js [hunk 18]
```diff
@@ -593,8 +600,7 @@ api.manageIdentity = brCallbackify(async ({actor, accountId, identityId}) => {
         httpStatusCode: 404,
         identity: identityId,
         account: accountId,
-        public: true
-      });
+        public: true});
```
Comment (dlongley) [inline lib/index.js]: Put this back

File: lib/index.js [hunk 19]
```diff
@@ -154,8 +154,8 @@ api.insert = brCallbackify(async ({actor, account, meta}) => {
   // validate resource roles (ensure actor is permitted to delegate the
   // roles specified in the account meta)
   if(meta.sysResourceRole.length > 0) {
-    await validateCapabilityDelegation(
-      {actor, resourceRoles: meta.sysResourceRole});
+    await validateCapabilityDelegation({
+      actor, resourceRoles: meta.sysResourceRole});
```
Comment (dlongley) [inline lib/index.js]: Put this back

File: lib/index.js (hunk 17 above)
Comment (aljones15) [reply lib/index.js]: clarify here.

Comment (dlongley) [reply lib/index.js]: The original format for this and the other linting changes were all correct, so we should change them back.

File: lib/index.js [hunk 20]
```diff
@@ -469,10 +479,8 @@ api.updateRoles = brCallbackify(async (
   // 2. add specified resource roles
   // 3. ensure resource roles are unique
   const resourceRoles = _.uniqWith(
-    brPermission.mergeCapabilities(
-      brPermission.subtractCapabilities(meta.sysResourceRole, remove),
-      add),
-    _.isEqual);
+    brPermission.mergeCapabilities(brPermission.subtractCapabilities(
+      meta.sysResourceRole, remove), add), _.isEqual);
```
Comment (dlongley) [inline lib/index.js]: Need this last one changed back -- it was much more readable before.

Comment (aljones15) [reply lib/index.js]: going to look in master and restore it to that.

**Comment by aljones15** [comment]
no issue really the linter thing is a bit annoying, but we need to solve it and no offense taken. I was actually way more concerned about how I prevent ids from being mutated.
